src/Parser/DeathKnight/Unholy/Modules/Items/ColdHeartEfficiency.js

```diff
@@ -21,49 +21,93 @@ class ColdHeartEfficiency extends Analyzer {
   totalColdHeartCasts = 0;
   correctColdHeartCasts = 0;
   buffColdHeart = 0;
+  
   unholyStrengthStart = 0;
   unholyStrengthRemaining = 0;
+  
+  concordanceStart = 0;
+  concordanceRemaining = 0;
+  
+  khazgorothStart = 0;
+  khazgorothRemaining = 0;
+  
+  timeAtMaxStacksStart = 0;
+  timeAtMaxStacks = 0;
 
   on_byPlayer_applybuff(event) {
     const spellID = event.ability.guid;
     if (spellID === SPELLS.UNHOLY_STRENGTH_BUFF.id) {
       this.unholyStrengthStart = event.timestamp;
     }
+	if (spellID === SPELLS.CONCORDANCE_OF_THE_LEGIONFALL_STRENGTH.id) {
+		this.concordanceStart = event.timestamp;
+	}
+	if (spellID === SPELLS.KHAZGOROTHS_SHAPING.id) {
+		this.khazgorothStart = event.timestamp;
+	}
   }
-
+  
+  on_byPlayer_refreshbuff(event) {
+	const spellID = event.ability.guid;
+	if (spellID === SPELLS.UNHOLY_STRENGTH_BUFF.id) {
+      this.unholyStrengthStart = event.timestamp;
+    }
+	if (spellID === SPELLS.CONCORDANCE_OF_THE_LEGIONFALL_STRENGTH.id) {
+		this.concordanceStart = event.timestamp;
+	}
+	if (spellID === SPELLS.KHAZGOROTHS_SHAPING.id) {
+		this.khazgorothStart = event.timestamp;
+	}
+  }
+  
   on_byPlayer_removebuff(event) {
     const spellID = event.ability.guid;
     if (spellID === SPELLS.UNHOLY_STRENGTH_BUFF.id) {
       this.unholyStrengthStart = 0;
     }
+	if (spellID === SPELLS.CONCORDANCE_OF_THE_LEGIONFALL_STRENGTH.id) {
+		this.concordanceStart = 0;
+    }
+	if (spellID === SPELLS.KHAZGOROTHS_SHAPING.id) {
+		this.khazgorothStart = 0;
+	}
   }
 
   on_byPlayer_applybuffstack(event) {
     const spellID = event.ability.guid;
     if (spellID === SPELLS.COLD_HEART_BUFF.id) {
       this.buffColdHeart = event.stack;
+	  if (this.buffColdHeart === 20) {
+		  this.timeAtMaxStacksStart = event.timestamp;
+	  }
     }
   }
 
   on_byPlayer_cast(event) {
     const spellId = event.ability.guid;
+	
     const unholyStrengthDuration = 15000;
+	const concordanceDuration = 10000;
+	const khazgorothDuration = 15000;
+	
     const coldHeartMaxStack = 20;
-    const coldHeartOptimalMinStack = 17;
-    const remainingUnholyStrengthAllowed = 4000;
+    const coldHeartOptimalMinStack = 14;
+	
+    const remainingDurationAllowed = 4000;
```

Comment:
The above consts should all be set as class constants at the top of this file, because they don't ever change.

I'm talking here about `remainingDurationAllowed` `coldHeartOptimalMinStack` `coldHeartMaxStack` `khazgorothDuration` `concordanceDuration` `unholyStrengthDuration`